File: python/tests/test_io.py
"""
Script: test_io.py
Objective: Verify Python raster loading and geometry validation.
Author: Yi Yu
Created: 2026-06-21
Last updated: 2026-06-21
Inputs: Bundled GeoTIFFs and in-memory RasterData objects.
Outputs: Pytest assertions for Python raster I/O behavior.
Usage: python3 -m pytest python/tests/test_io.py -v
Dependencies: numpy, pytest, rasterio, and ubestarfm.
"""

from __future__ import annotations

import numpy as np
import pytest
from rasterio.crs import CRS
from rasterio.transform import from_origin

from ubestarfm.io import RasterData, read_raster, validate_geometry


def test_read_raster_returns_float64_row_major_array(example_path) -> None:
    raster = read_raster(example_path("MOD11A1_LST_cloudrm_20160218.tif"))

    assert raster.values.shape == (400, 400)
    assert raster.values.dtype == np.float64
    assert raster.values.flags.c_contiguous
    assert raster.values[0, 0] == pytest.approx(312.74, abs=1e-4)
    assert np.isfinite(raster.values).all()


def test_geometry_mismatch_is_rejected() -> None:
    first = RasterData(
        values=np.ones((3, 3), dtype=np.float64),
        profile={
            "height": 3,
            "width": 3,
            "transform": from_origin(0, 3, 1, 1),
            "crs": CRS.from_epsg(4326),
            "nodata": np.nan,
            "dtype": "float64",
            "count": 1,
            "driver": "GTiff",
        },
    )
    second = RasterData(
        values=np.ones((4, 3), dtype=np.float64),
        profile={
            **first.profile,
            "height": 4,
            "transform": from_origin(0, 4, 1, 1),
        },
    )

    with pytest.raises(ValueError, match="same dimensions, transform, and CRS"):
        validate_geometry([first, second])


def test_geometry_accepts_floating_point_equivalent_transforms() -> None:
    first = RasterData(
        values=np.ones((2, 2), dtype=np.float64),
        profile={
            "height": 2,
            "width": 2,
            "transform": from_origin(146, -34.8, 0.001, 0.001),
            "crs": CRS.from_epsg(4326),
            "nodata": np.nan,
            "dtype": "float64",
            "count": 1,
            "driver": "GTiff",
        },
    )
    second = RasterData(
        values=np.ones((2, 2), dtype=np.float64),
        profile={
            **first.profile,
            "transform": from_origin(
                146,
                -34.8,
                0.001,
                0.001 + 2e-17,
            ),
        },
    )

    validate_geometry([first, second])

File: tests/testthat/test-io.R
# Script: test-io.R
# Objective: Verify R raster loading and geometry validation.
# Author: Yi Yu
# Created: 2026-06-21
# Last updated: 2026-06-21
# Inputs: Bundled GeoTIFF fixtures and in-memory SpatRaster objects.
# Outputs: testthat assertions for raster I/O behavior.
# Usage: Rscript -e 'testthat::test_file("tests/testthat/test-io.R")'
# Dependencies: R packages testthat, terra, and ubestarfm.

test_that("single-layer GeoTIFFs load as row-column matrices", {
  input <- ubestarfm_read_raster(
    example_path("MOD11A1_LST_cloudrm_20160218.tif")
  )

  expect_s4_class(input$raster, "SpatRaster")
  expect_equal(dim(input$values), c(400L, 400L))
  expect_equal(input$values[1, 1], 312.74, tolerance = 1e-4)
  expect_true(all(is.finite(input$values)))
})

test_that("reference rasters must share geometry", {
  first <- terra::rast(
    nrows = 3,
    ncols = 3,
    xmin = 0,
    xmax = 3,
    ymin = 0,
    ymax = 3
  )
  second <- terra::rast(
    nrows = 4,
    ncols = 3,
    xmin = 0,
    xmax = 3,
    ymin = 0,
    ymax = 4
  )

  expect_error(
    ubestarfm_validate_geometry(list(first, second)),
    "same dimensions, extent, resolution, and CRS"
  )
})
